File: eco-core/core-plugin/src/main/java/com/willfp/ecobosses/commands/CommandSpawn.java
package com.willfp.ecobosses.commands;

import com.willfp.eco.core.command.impl.Subcommand;
import com.willfp.eco.core.config.updating.ConfigUpdater;
import com.willfp.ecobosses.EcoBossesPlugin;
import com.willfp.ecobosses.bosses.EcoBoss;
import com.willfp.ecobosses.bosses.EcoBosses;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.util.StringUtil;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class CommandSpawn extends Subcommand {
    /**
     * The cached names.
     */
    private static final List<String> BOSS_NAMES = new ArrayList<>();

    /**
     * The cached numbers.
     */
    private static final List<String> TILDE = Arrays.asList(
            "~"
    );

    /**
     * Instantiate a new executor for /ebspawn.
     *
     * @param plugin The plugin to manage the execution for.
     */
    public CommandSpawn(@NotNull final EcoBossesPlugin plugin) {
        super(plugin, "spawn", "ecobosses.command.spawn", false);
        reload();
    }

    /**
     * Called on reload.
     */
    @ConfigUpdater
    public static void reload() {
        BOSS_NAMES.clear();
        BOSS_NAMES.addAll(EcoBosses.values().stream().map(EcoBoss::getId).collect(Collectors.toList()));
    }

    @Override
    public void onExecute(@NotNull final CommandSender sender,
                          @NotNull final List<String> args) {
        if (args.isEmpty()) {
            sender.sendMessage(this.getPlugin().getLangYml().getMessage("specify-boss"));
            return;
        }

        String bossName = args.get(0);

        EcoBoss boss = EcoBosses.getByName(bossName.toLowerCase());

        if (boss == null) {
            sender.sendMessage(this.getPlugin().getLangYml().getMessage("specify-boss"));
            return;
        }

        Location location = null;

        if (sender instanceof Player) {
            location = ((Player) sender).getLocation();
        }

        if (args.size() >= 4) {
            String xString = args.get(1);
            String yString = args.get(2);
            String zString = args.get(3);

            double xPos;
            double yPos;
            double zPos;

            if (xString.startsWith("~") && sender instanceof Player) {
                String xDiff = xString.replace("~", "");
                String yDiff = yString.replace("~", "");
                String zDiff = zString.replace("~", "");

                if (xDiff.isEmpty()) {
                    xPos = ((Player) sender).getLocation().getX();
                } else {
                    try {
                        xPos = ((Player) sender).getLocation().getX() + Double.parseDouble(xDiff);
                    } catch (NumberFormatException e) {
                        xPos = ((Player) sender).getLocation().getX();
                    }
                }

                if (yDiff.isEmpty()) {
                    yPos = ((Player) sender).getLocation().getY();
                } else {
                    try {
                        yPos = ((Player) sender).getLocation().getY() + Double.parseDouble(yDiff);
                    } catch (NumberFormatException e) {
                        yPos = ((Player) sender).getLocation().getY();
                    }
                }

                if (zDiff.isEmpty()) {
                    zPos = ((Player) sender).getLocation().getZ();
                } else {
                    try {
                        zPos = ((Player) sender).getLocation().getZ() + Double.parseDouble(yDiff);
                    } catch (NumberFormatException e) {
                        zPos = ((Player) sender).getLocation().getZ();
                    }
                }
            } else {
                try {
                    xPos = Double.parseDouble(xString);
                    yPos = Double.parseDouble(yString);
                    zPos = Double.parseDouble(zString);
                } catch (NumberFormatException e) {
                    sender.sendMessage(this.getPlugin().getLangYml().getMessage("invalid-location"));
                    return;
                }
            }

            location = new Location(null, xPos, yPos, zPos);
        }

        World world = null;
        if (sender instanceof Player) {
            world = ((Player) sender).getWorld();
        }

        if (args.size() >= 5) {
            world = Bukkit.getWorld(args.get(4));
        }

        if (location == null) {
            sender.sendMessage(this.getPlugin().getLangYml().getMessage("invalid-location"));
            return;
        }

        location.setWorld(world);

        if (world == null) {
            sender.sendMessage(this.getPlugin().getLangYml().getMessage("invalid-world"));
            return;
        }

        boss.spawn(location);
        sender.sendMessage(this.getPlugin().getLangYml().getMessage("spawned"));
    }

    @Override
    public List<String> tabComplete(@NotNull final CommandSender sender,
                                    @NotNull final List<String> args) {
        List<String> completions = new ArrayList<>();

        if (args.isEmpty()) {
            // Currently, this case is not ever reached
            return new ArrayList<>();
        }

        if (args.size() == 1) {
            StringUtil.copyPartialMatches(args.get(0), BOSS_NAMES, completions);

            Collections.sort(completions);
            return completions;
        }

        if (args.size() == 2) {
            StringUtil.copyPartialMatches(args.get(1), TILDE, completions);

            Collections.sort(completions);
            return completions;
        }

        if (args.size() == 3) {
            StringUtil.copyPartialMatches(args.get(2), TILDE, completions);

            Collections.sort(completions);
            return completions;
        }

        if (args.size() == 4) {
            StringUtil.copyPartialMatches(args.get(3), TILDE, completions);

            Collections.sort(completions);
            return completions;
        }

        if (args.size() == 5) {
            StringUtil.copyPartialMatches(args.get(4), Bukkit.getWorlds().stream().map(World::getName).collect(Collectors.toList()), completions);

            Collections.sort(completions);
            return completions;
        }

        return new ArrayList<>(0);
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/ecobosses/commands/CommandSpawn.kt
package com.willfp.ecobosses.commands

import com.willfp.eco.core.command.impl.Subcommand
import com.willfp.ecobosses.EcoBossesPlugin
import com.willfp.ecobosses.bosses.Bosses
import com.willfp.ecobosses.events.BossSpawnEvent
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.World
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import org.bukkit.util.StringUtil

class CommandSpawn(plugin: EcoBossesPlugin) : Subcommand(
    plugin,
    "spawn",
    "ecobosses.command.spawn",
    false
) {
    override fun onExecute(
        sender: CommandSender,
        args: List<String>
    ) {
        if (args.isEmpty()) {
            sender.sendMessage(plugin.langYml.getMessage("specify-boss"))
            return
        }

        val bossName = args[0]
        val boss = Bosses.getByID(bossName.lowercase())
        if (boss == null) {
            sender.sendMessage(plugin.langYml.getMessage("specify-boss"))
            return
        }

        var location: Location? = null
        if (sender is Player) {
            location = sender.location
        }

        if (args.size >= 4) {
            val xString = args[1]
            val yString = args[2]
            val zString = args[3]
            val xPos: Double
            val yPos: Double
            val zPos: Double
            if (xString.startsWith("~") && sender is Player) {
                val xDiff = xString.replace("~", "")
                val yDiff = yString.replace("~", "")
                val zDiff = zString.replace("~", "")
                xPos = if (xDiff.isEmpty()) {
                    sender.location.x
                } else {
                    try {
                        sender.location.x + xDiff.toDouble()
                    } catch (e: NumberFormatException) {
                        sender.location.x
                    }
                }
                yPos = if (yDiff.isEmpty()) {
                    sender.location.y
                } else {
                    try {
                        sender.location.y + yDiff.toDouble()
                    } catch (e: NumberFormatException) {
                        sender.location.y
                    }
                }
                zPos = if (zDiff.isEmpty()) {
                    sender.location.z
                } else {
                    try {
                        sender.location.z + yDiff.toDouble()
                    } catch (e: NumberFormatException) {
                        sender.location.z
                    }
                }
            } else {
                try {
                    xPos = xString.toDouble()
                    yPos = yString.toDouble()
                    zPos = zString.toDouble()
                } catch (e: NumberFormatException) {
                    sender.sendMessage(plugin.langYml.getMessage("invalid-location"))
                    return
                }
            }
            location = Location(null, xPos, yPos, zPos)
        }
        var world: World? = null
        if (sender is Player) {
            world = sender.world
        }
        if (args.size >= 5) {
            world = Bukkit.getWorld(args[4])
        }
        if (location == null) {
            sender.sendMessage(plugin.langYml.getMessage("invalid-location"))
            return
        }

        location.world = world

        if (world == null) {
            sender.sendMessage(plugin.langYml.getMessage("invalid-world"))
            return
        }

        val event = BossSpawnEvent(
            boss,
            location,
            BossSpawnEvent.SpawnReason.COMMAND,
            null
        )

        Bukkit.getPluginManager().callEvent(event)

        if (!event.isCancelled) {
            boss.spawn(location)
        }

        sender.sendMessage(plugin.langYml.getMessage("spawned"))
    }

    override fun tabComplete(
        sender: CommandSender,
        args: List<String>
    ): List<String> {
        val completions = mutableListOf<String>()

        if (args.isEmpty()) {
            return emptyList()
        }

        if (args.size == 1) {
            StringUtil.copyPartialMatches(args[0], BOSS_NAMES, completions)
            completions.sort()
            return completions
        }
        if (args.size == 2) {
            StringUtil.copyPartialMatches(args[1], TILDE, completions)
            completions.sort()
            return completions
        }
        if (args.size == 3) {
            StringUtil.copyPartialMatches(args[2], TILDE, completions)
            completions.sort()
            return completions
        }
        if (args.size == 4) {
            StringUtil.copyPartialMatches(args[3], TILDE, completions)
            completions.sort()
            return completions
        }
        if (args.size == 5) {
            StringUtil.copyPartialMatches(args[4], Bukkit.getWorlds().map { it.name }, completions)
            completions.sort()
            return completions
        }
        return ArrayList(0)
    }

    companion object {
        /**
         * The cached names.
         */
        private val BOSS_NAMES: List<String>
            get() = Bosses.values().map { it.id }

        /**
         * The cached numbers.
         */
        private val TILDE = listOf(
            "~"
        )
    }
}